src/main/java/org/commcare/core/interfaces/HttpResponseProcessor.java

```diff
@@ -12,7 +12,7 @@ public interface HttpResponseProcessor {
     /**
      * Http response was in the 200s
      */
-    void processSuccess(int responseCode, InputStream responseData);
+    void processSuccess(int responseCode, InputStream responseData, String apiVersion);
```

Comment:
is `apiVersion` still required here ?

Reply:
This is how I planned to expose the value to the calling code. Specifically, after gets/posts in ConnectNetworkHelper I plan to pass the expected version into the getResponseProcessor call, and check it against the received value included in the processSuccess call [here](https://github.com/dimagi/commcare-android/blob/e91160f5f15f4009ebfa3e89d967e630371fd691/app/src/org/commcare/connect/network/ConnectNetworkHelper.java#L341). Does that sound like the right approach?

Reply:
I see, I don't think we need to expose it specifically in `processSuccess`, instead implementations of `processSucess` can directly use the `responseData` to get `apiVersion` ?

Also I am confused if `processSuccess` is where callers will be handling `apiVersion` why are we adding `public String getApiVersion() {` in here and ModernHTTPTask ?

Reply:
Maybe a better way to phrase my concern with the second approach: If ModernHttpTask is a ResponseStreamAccessor that allows us access to the InputStream and API string, shouldn't we be accessing both the same way?

So if we're going to pass the InputStream via processSuccess instead of calling myModernHttpTask.getResponseStream, I think the API string should be provided the same way.

Reply:
> That way access to the response stream and API version would only be available in the callback (I'm going with the proposal to have processSuccess pass the ResponseStreamAccessor). In other words, it wouldn't be possible for calling code to attempt to access the response stream or API string from the task/requester before actually executing the request. 
 
This sounds like a good design choice to me although 

> perhaps instead of implementing ResponseStreamAccessor, both ModernHttpRequester and ModernHttpTask should create and return generic implementations of that interface when calling processResponse.

I am not super clear what you mean by this and having some pseudo code to look at might be helpful.

Reply:
No problem. The generic ResponseStreamAccessor would be created when calling processResponse in [ModernHttpTask ](https://github.com/dimagi/commcare-android/blob/776435132ba9f0ad9f3bcca41943fa2ee896c45e/app/src/org/commcare/tasks/ModernHttpTask.java#L90)and [ModernHttpRequester](https://github.com/dimagi/commcare-core/blob/7c8dde6ac6e602468d1802328d75ef0d2faf095a/src/main/java/org/commcare/core/network/ModernHttpRequester.java#L90), rather than returning `this` as the last function input.

For ModernHttpTask, the new code would look like:
![image](https://github.com/dimagi/commcare-core/assets/1054894/75adc485-aca1-4237-bbe3-1a1e0cbad8e3)

The class would no longer implement `ResponseStreamAccessor`, and the `getResponseStream` and `getApiVersion` methods at the end of the file would be deleted.

And in places like [this](https://github.com/dimagi/commcare-android/blob/776435132ba9f0ad9f3bcca41943fa2ee896c45e/app/src/org/commcare/activities/PostRequestActivity.java#L120C17-L120C25), it would no longer be possible to call `postTask.getResponseStream` before calling `connect` and `executeParallel`. Note that I currently don't see this happening anywhere in the code, but the potential is there.

Reply:
yes, that seems promising to me 👍

Reply:
Implemented these changes